src/graph.rs

```diff
@@ -130,37 +140,28 @@ impl Range {
 
 #[derive(Debug, Clone)]
 pub enum ModuleGraphError {
-  InvalidTypeAssertion {
-    specifier: ModuleSpecifier,
-    range: Range,
-    actual_media_type: MediaType,
-    expected_media_type: MediaType,
-  },
   LoadingErr(ModuleSpecifier, Option<Range>, Arc<anyhow::Error>),
   Missing(ModuleSpecifier, Option<Range>),
   ParseErr(ModuleSpecifier, deno_ast::Diagnostic),
-  ResolutionError(ResolutionError),
-  UnsupportedImportAssertionType {
-    specifier: ModuleSpecifier,
-    range: Range,
-    kind: String,
-  },
   UnsupportedMediaType {
     specifier: ModuleSpecifier,
     media_type: MediaType,
     maybe_referrer: Option<Range>,
   },
+  // Note: Resolution and import errors are special in that they don't occupy
+  // module slots, whereas the other variants do.
+  ResolutionError(ResolutionError),
+  ImportError(ImportError),
 }
 
 impl ModuleGraphError {
   pub fn specifier(&self) -> &ModuleSpecifier {
     match self {
+      Self::ImportError(_) => panic!("Import errors don't have specifiers"),
```

Comment:
We should remove this panic. Either all the errors should have a specifier or this should become `Option<&ModuleSpecifier>`. I'm not sure what's changed that there wouldn't be a specifier anymore though?

Reply:
> I'm not sure what's changed that there wouldn't be a specifier anymore though?

Yeah I've been meaning to complain about this method... the specifier returned by `ResolutionError` has a drastically different meaning than the other variants (referrer vs target). Similar with `maybe_range()`. `ImportError` kind of aligns with `ResolutionError` on that.

It's a symptom of `ModuleGraphError` mixing errors that can occupy module slots vs errors that represent arrows between them. A year ago on a previous attempt at this fix I tried separating them more thoroughly but the boilerplate was extreme. So yeah I think we should just make it return an `Option`.